design/history/exploration-reports/2020.09-learning-ipld.md

```diff
@@ -0,0 +1,98 @@
+were relatively easy to understand.
+
+### First impression
+
+The docs are somewhat scattered and unfinished, which does make it a little
+extra confusing to get started. In chronological order, I read:
+
+1) https://hackmd.io/LHTTmGSWSvem4Wz2h_a39g?both, Eric's "terse primer". Seems
+   to try to cover everything, though it does seem like quite a lot of
+   information to take in all at once.
+
+2) https://ipld.github.io/docs/, sourced at https://github.com/ipld/docs. Seems
+   aimed at getting started with tutorials in JS. I should note that I first
+   read this before Mikeal's "new intro" added on September 2nd 2020.
+
+3) https://github.com/ipld/specs, which seems to contain all the formal specs,
+   but also includes a pretty decent README.
+
+I think all three should probably be unified into two halves:
+
+* A high-level introduction to IPLD, max 3-4 pages. Probably extending Mikeal's
+  new intro with some material from Eric's primer?
+
+* The set of spec documents, with a README to classify and introduce each of the
+  groups or layers. I think the specs repo already does a decent job at this.
+
+### Concepts that confused me
+
+I already raised some of these as Slack threads or HackMD comments, but for the
+sake of keeping record, I'm listing the most basic or important ones here.
+
+* It is said that a link is unlike a URL, since it is merely a hash of data
+  that doesn't statically say where to fetch the data from. So... how would one
+  ever actually fetch data via a link?
+
+* Out of the three layers (blocks, data model, schemas), Schemas have been by
+  far the hardest to wrap my head around. I think an introduction should contain
+  a very brief example, including how it actually looks like when mapped to the
+  data model and encoded into a block.
+
+The following are more such points, but focused around schemas, once I got to
+that part of the spec:
+
+* My first read about ADLs left me very confused, in particular how they're
+  different than Schemas. I found the "Mapped to the Data Model" introduction to
+  ADLs much easier to understand, as it shows reasonable examples.
+
+* Why are multi-block data structures a separate definition in the spec, and not
+  just part of Schemas?
+
+* Are blocks generally filled with data nearly completely, or is it normal to
+  have them relatively empty?
+
+* Wouldn't removing the first byte from a very long multi-block List mean that
+  every block would need to be modified to shift all bytes forward by one? I
+  assume and hope not, but the spec doesn't really give pointers.
+
+* Since data in blocks is encoded from the data model, how would Iknow if a
+  particular data model value fits in a single block? What about a shema value?
+
+* Would IPLD be much different if the data model was an internal detail, and not
+  exposed to the user? I imagine that, most of the time, one would interact with
+  schemas and not the data model.
```

Comment:
Since these are all just protocols you have to keep in mind how a-la-carte people will choose to adopt them. Data encoded by one actor using a schema may be decoded by another actor who has little interest in the schema and can accomplish what they want with just the encoded data, and they may not even implement a full featured deserializer in order to do that.